lib/edge_impulse/ei_ncs.cpp

```diff
@@ -0,0 +1,404 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <assert.h>
+#include <ei_run_classifier.h>
+#include <ei_ncs.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ei_ncs, CONFIG_EI_NCS_LOG_LEVEL);
+
+#define INPUT_FRAME_SIZE	EI_CLASSIFIER_RAW_SAMPLES_PER_FRAME
+#define INPUT_WINDOW_SIZE	EI_CLASSIFIER_DSP_INPUT_FRAME_SIZE
+
+#define PROCESS_BUF		1
+BUILD_ASSERT(PROCESS_BUF > 0);
+
+BUILD_ASSERT(CONFIG_EI_NCS_THREAD_STACK_SIZE > 0);
+
+#define DATA_BUFFER_SIZE	CONFIG_EI_NCS_DATA_BUF_SIZE
+#define THREAD_STACK_SIZE	CONFIG_EI_NCS_THREAD_STACK_SIZE
+#define THREAD_PRIORITY 	CONFIG_EI_NCS_THREAD_PRIORITY
+#define DEBUG_MODE		IS_ENABLED(CONFIG_EI_NCS_DEBUG_MODE)
+
+enum state {
+	STATE_DISABLED,
+	STATE_PROCESSING,
+	STATE_READY,
+};
+
+struct data_buffer {
+	float buf[DATA_BUFFER_SIZE];
+	float *processing_head;
+	float *append_head;
+	size_t wait_data_size;
+	struct k_spinlock lock;
+	enum state state;
+};
+
+static K_THREAD_STACK_DEFINE(thread_stack, THREAD_STACK_SIZE);
+static struct k_thread thread;
+static k_tid_t ei_thread_id;
+
+static K_SEM_DEFINE(ei_sem, 0, 1);
+
+static struct data_buffer ei_input;
+static ei_impulse_result_t ei_result;
+static ei_ncs_result_ready_cb user_cb;
+
+
+BUILD_ASSERT(DATA_BUFFER_SIZE > INPUT_WINDOW_SIZE);
+BUILD_ASSERT(INPUT_WINDOW_SIZE % INPUT_FRAME_SIZE == 0);
+
+
+static size_t buf_calc_max_processing_move(struct data_buffer *b)
+{
+	size_t res;
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	if (b->append_head >= b->processing_head) {
+		res = b->append_head - b->processing_head;
+	} else {
+		res = (buf_end - b->processing_head) +
+		      (b->append_head - buf_start);
+	}
+
+	return res;
+}
+
+static size_t buf_calc_free_space(struct data_buffer *b)
+{
+	size_t res;
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	if (b->wait_data_size > 0) {
+		res = b->wait_data_size + ARRAY_SIZE(b->buf) -
```

Comment:
res is not needed - see comment at line 64

general comment - avoid unnecessary indent
```
if (a) {
} else if (b) {
} else {
   // c
}
```

In this case you can (almost) do the following:
```
if (wait_data_size > 0) {
   return wait_size_case;
}
return ??? + buf_calc_max_processing_move(b) - 1;
```
Is it possible to find something that would utilize the max move?

Reply:
We could separate calculating difference between pointers that are used for the circular buffer.